src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs

```diff
@@ -38,7 +40,12 @@ public sealed override TypeWithAnnotations TypeWithAnnotations
         {
             get
             {
-                return this.RetargetingModule.RetargetingTranslator.Retarget(_underlyingParameter.TypeWithAnnotations, RetargetOptions.RetargetPrimitiveTypesByTypeCode);
+                if (!_lazyTypeWithAnnotations.HasValue)
+                {
+                    _lazyTypeWithAnnotations = this.RetargetingModule.RetargetingTranslator.Retarget(_underlyingParameter.TypeWithAnnotations, RetargetOptions.RetargetPrimitiveTypesByTypeCode);
```

Comment:
`Nullable<TypeWithAnnotations>` is not thread-safe. Is this field accessed from multiple threads?

Reply:
@cston Yeah I was chatting with Sam about this. I didn't find an appropriate existing helper, so Sam is working on merging helpers that are not available in the compilers layer with those available in the compilers layer.

Reply:
No additional helpers are necessary, that's what https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/TypeWithAnnotations.cs,e5078e3d5fd03b05,references is for.

Reply:
If the goal is to avoid recalculating the type, it may be sufficient to just cache the `TypeSymbol`.
```csharp
private TypeSymbol? _lazyType;

public sealed override TypeWithAnnotations TypeWithAnnotations
{
    get
    {
        var underlyingType = _underlyingParameter.TypeWithAnnotations;
        var translator = RetargetingModule.RetargetingTranslator;
        if (_lazyType is null)
        {
            Interlocked.CompareExchange(ref _lazyType,
                translator.Retarget(underlyingType.Type,
                    RetargetOptions.RetargetPrimitiveTypesByTypeCode),
                null);
        }
        var newModifiers = translator.RetargetModifiers(underlyingType.CustomModifiers, out _);
        return underlyingType.WithTypeAndModifiers(_lazyType, newModifiers);
    }
}
```

Reply:
@cston In the trace I provided, this property is accessed 12,833,732 times during 3 minutes. So I think it's worth to just cache the whole thing.

Reply:
> No additional helpers are necessary, that's what https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/TypeWithAnnotations.cs,e5078e3d5fd03b05,references is for.

Does this suggest it like that?

```csharp
if (_lazyTypeWithAnnotations is null)
{
    Interlocked.CompareExchange(ref _lazyTypeWithAnnotations, new TypeWithAnnotations.Boxed(......), null);
}
```

I think that would still risk calculating it twice if two threads concurrently accessed the property.

What Sam suggested was something to be like this:

```csharp
return LazyInitialization.EnsureInitialized(
    ref _lazyTypeWithAnnotations,
    static self => self.RetargetingModule.RetargetingTranslator.Retarget(self._underlyingParameter.TypeWithAnnotations, RetargetOptions.RetargetPrimitiveTypesByTypeCode),
    this);
```

Reply:
@333fred The helpers are lazy initialization helpers for reference types, and are unrelated to the fact that this is a value type.

Reply:
We should switch to ```TypeWithAnnotations.Boxed```, which we use to cache this information in symbols. There should be many examples. No special helpers are necessary and the change shouldn't be blocked on that.